src/clients/kol.ts

```diff
@@ -87,95 +114,239 @@ function sanitiseBlueText(blueText: string | undefined): string {
+          data: this._loginParameters,
+          maxRedirects: 0,
+          validateStatus: (status) => status === 302,
+        });
+        const sessionCookies = (loginResponse.headers["set-cookie"] || [])
+          .map((cookie: string) => cookie.split(";")[0])
+          .join("; ");
+        const apiResponse = await axios("https://www.kingdomofloathing.com/api.php", {
+          withCredentials: true,
+          headers: {
+            cookie: sessionCookies,
+          },
+          params: {
+            what: "status",
+            for: `${this._loginParameters.get("loginname")} Chatbot`,
+          },
+        });
+        this._credentials = {
+          sessionCookies: sessionCookies,
+          pwdhash: apiResponse.data.pwd,
+        };
+        return true;
+      } catch {
+        console.log("Login failed. Checking if it's because of rollover.");
+        await this.rolloverCheck();
+        return false;
       }
     });
   }
 
-  private async makeCredentialedRequest(url: string, parameters: object) {
+  private chatBotStarted = false;
+
+  async startChatBot() {
+    if (this.chatBotStarted) return;
+    this.loopChatBot();
+    this.chatBotStarted = true;
+  }
+
+  private async loopChatBot() {
+    await Promise.all([this.checkWhispers(), this.checkKmails()]);
+    await wait(3000);
+    await this.loopChatBot();
+  }
+
+  private lastFetchedWhispers = "0";
+
+  async checkWhispers() {
+    const newChatMessagesResponse = (await this.visitUrl("newchatmessages.php", {
+      j: 1,
+      lasttime: this.lastFetchedWhispers,
+    })) as { last: string; msgs: KoLChatMessage[] };
+
+    if (!newChatMessagesResponse || typeof newChatMessagesResponse !== "object") return;
+
+    this.lastFetchedWhispers = newChatMessagesResponse["last"];
+
+    newChatMessagesResponse["msgs"]
+      .filter((msg) => msg["type"] === "private" && msg.who && msg.msg)
+      .map((msg) => ({
+        who: {
+          id: Number(msg.who!.id),
+          name: msg.who!.name,
+        },
+        msg: msg.msg!,
+        time: new Date(Number(msg.time) * 1000),
+      }))
+      .forEach((whisper) => this.emit("whisper", whisper));
+  }
+
+  async checkKmails() {
+    const newKmailsResponse = (await this.visitUrl("api.php", {
+      what: "kmail",
+      for: `${this._loginParameters.get("loginname")} Chatbot`,
+    })) as KoLKmail[];
+
+    if (!Array.isArray(newKmailsResponse) || newKmailsResponse.length === 0) return;
+
+    const newKmails = newKmailsResponse.map((msg: KoLKmail) => ({
+      who: {
+        id: Number(msg.fromid),
+        name: msg.fromname,
+      },
+      msg: msg.message,
+      time: new Date(Number(msg.azunixtime) * 1000),
+    }));
+
+    const data = {
+      the_action: "delete",
+      pwd: this._credentials?.pwdhash,
+      box: "Inbox",
+      ...newKmailsResponse.reduce(
+        (acc: { [x: string]: string }, msg: KoLKmail) => ({
+          ...acc,
+          [`sel${msg.id}`]: "on",
+        }),
+        {}
```

Comment:
```suggestion
      ...Object.fromEntries(newKmailsResponse.map(({ id }) => [`sel${id}`, "ojn"])),
```

Reply:
ojn

Reply:
https://en.wikipedia.org/wiki/Order_of_Julian_of_Norwich
my best friend wrote a thesis on abject imagery in Julian of Norwich's writings, and drew the obvious comparison to Serrano's _Piss Christ_

Reply:
this is in general an extremely optional recommendation it just felt right to me